bundle-workflow/python/build_workflow/builder.py

```diff
@@ -0,0 +1,48 @@
+# Copyright OpenSearch Contributors.
+# SPDX-License-Identifier: Apache-2.0
+
+import os
+
+'''
+This class is responsible for executing the build for a component and passing the results to a build recorder.
+It will notify the build recorder of build information such as repository and git ref, and any artifacts generated by the build.
+Artifacts found in "<build root>/artifacts/<maven|plugins|libs|bundle>" will be recognized and recorded.
+'''
+class Builder:
+    def __init__(self, component_name, git_repo, component_scripts_path, default_build_script, build_recorder):
+        '''
+        Construct a new Builder instance.
+        :param component_name: The name of the component to build.
+        :param git_repo: A GitRepository instance containing the checked-out code.
+        :param component_scripts_path: Where to look for individual component "build.sh" scripts for those repositories that don't contain their own "scripts/build.sh".
+        :param default_build_script: Path to a defauld build script to use when neither a per-repo "scripts/build.sh" nor a "component_scripts_path/<component>/build.sh" script can be found.
+        :param build_recorder: The build recorder that will capture build information and artifacts.
+        '''
+
+        self.component_name = component_name
+        self.git_repo = git_repo
+        self.component_scripts_path = component_scripts_path
+        self.default_build_script = default_build_script
+        self.build_recorder = build_recorder
+
+    def build(self, version, arch):
+        paths = [os.path.realpath(os.path.join(self.git_repo.dir.name, 'scripts/build.sh')),
```

Comment:
I would extract this into a `build_script` method. We don't write tests for this code, but if we did we'd want to make sure we explain the logic of selection of the build script.

There's also a smell here that we have 3 resolutions of a build path in what seems like the wrong place: one in the checked out code (belongs with the component), one in the override location (belongs with the component builder), and the default one (belongs in the bundle builder).